File: java-service/src/main/java/com/algoverse/Algorithms.java
package com.algoverse;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Algorithms {

    public static class Metadata {
        public Integer line_number;
        public Metadata(Integer line_number) {
            this.line_number = line_number;
        }
    }

    public static class Event {
        public List<Integer> values;
        public List<Integer> indices;
        public String category;
        public String event;
        public String description;
        public Metadata metadata;
        public long timestamp;

        public Event(int[] array, List<Integer> highlighted, String category, String event, String description, Integer lineNumber) {
            if (array != null) {
                this.values = new ArrayList<>();
                for (int val : array) {
                    this.values.add(val);
                }
            } else {
                this.values = null;
            }
            this.indices = highlighted != null ? highlighted : new ArrayList<>();
            this.category = category;
            this.event = event;
            this.description = description;
            if (lineNumber != null) {
                this.metadata = new Metadata(lineNumber);
            }
            this.timestamp = System.currentTimeMillis();
        }
    }

    public static class Metrics {
        public double time_ms;
        public Object comparisons = null;
        public Object swaps = null;

        public Metrics(double time_ms) {
            this.time_ms = time_ms;
        }
    }

    public static class Complexity {
        public String time;
        public String space;

        public Complexity(String time, String space) {
            this.time = time;
            this.space = space;
        }
    }

    public static class AlgorithmResult {
        public String version = "1.0";
        public String language = "java";
        public String algorithm;
        public String mode;
        public List<Event> events;
        public Metrics metrics;
        public Complexity complexity;

        public AlgorithmResult(String algorithm, String mode, List<Event> events, double time_ms) {
            this.algorithm = algorithm;
            this.mode = mode;
            this.events = events;
            this.metrics = new Metrics(time_ms);
            
            if (algorithm.equals("bubble_sort") || algorithm.equals("insertion_sort") || algorithm.equals("selection_sort") || algorithm.equals("shell_sort")) {
                this.complexity = new Complexity("O(n^2)", "O(1)");
            } else if (algorithm.equals("quick_sort") || algorithm.equals("merge_sort") || algorithm.equals("heap_sort")) {
                this.complexity = new Complexity("O(n log n)", algorithm.equals("merge_sort") ? "O(n)" : "O(log n)");
            } else if (algorithm.equals("radix_sort")) {
                this.complexity = new Complexity("O(nk)", "O(n+k)");
            } else {
                this.complexity = new Complexity("unknown", "unknown");
            }
        }
    }

    private static void pushEvent(List<Event> events, int[] arr, List<Integer> highlighted, String category, String event, String description, Integer lineNumber, boolean visualization) {
        if (visualization) {
            events.add(new Event(arr, highlighted, category, event, description, lineNumber));
        }
    }

    public static AlgorithmResult runBubbleSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Bubble Sort", 1, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                pushEvent(events, arr, Arrays.asList(j, j + 1), "comparison", "compare", "Comparing elements", 5, visualization);
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                    pushEvent(events, arr, Arrays.asList(j, j + 1), "array_mutation", "swap", "Swapped elements", 8, visualization);
                }
            }
            if (!swapped) break;
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Bubble Sort complete", 12, visualization);
        return new AlgorithmResult("bubble_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    public static AlgorithmResult runQuickSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Quick Sort", 18, visualization);
        long start = System.nanoTime();
        
        quickSortHelper(arr, 0, arr.length - 1, events, visualization);
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Quick Sort complete", 24, visualization);
        return new AlgorithmResult("quick_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    private static void quickSortHelper(int[] arr, int low, int high, List<Event> events, boolean visualization) {
        if (low < high) {
            int pi = partition(arr, low, high, events, visualization);
            quickSortHelper(arr, low, pi - 1, events, visualization);
            quickSortHelper(arr, pi + 1, high, events, visualization);
        }
    }

    private static int partition(int[] arr, int low, int high, List<Event> events, boolean visualization) {
        int pivot = arr[high];
        pushEvent(events, arr, Arrays.asList(high), "comparison", "pivot", "Chosen pivot", 2, visualization);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            pushEvent(events, arr, Arrays.asList(j, high), "comparison", "compare", "Comparing with pivot", 5, visualization);
            if (arr[j] < pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                pushEvent(events, arr, Arrays.asList(i, j), "array_mutation", "swap", "Swapped elements", 9, visualization);
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        pushEvent(events, arr, Arrays.asList(i + 1, high), "array_mutation", "partition", "Pivot placed in correct position", 14, visualization);
        return i + 1;
    }

    public static AlgorithmResult runMergeSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Merge Sort", 17, visualization);
        long start = System.nanoTime();
        
        mergeSortHelper(arr, 0, arr.length - 1, events, visualization);
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Merge Sort complete", 24, visualization);
        return new AlgorithmResult("merge_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    private static void mergeSortHelper(int[] arr, int left, int right, List<Event> events, boolean visualization) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSortHelper(arr, left, mid, events, visualization);
            mergeSortHelper(arr, mid + 1, right, events, visualization);
            merge(arr, left, mid, right, events, visualization);
        }
    }

    private static void merge(int[] arr, int left, int mid, int right, List<Event> events, boolean visualization) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        int[] L = new int[n1];
        int[] R = new int[n2];
        
        for (int i = 0; i < n1; ++i) L[i] = arr[left + i];
        for (int j = 0; j < n2; ++j) R[j] = arr[mid + 1 + j];
        
        int i = 0, j = 0;
        int k = left;
        while (i < n1 && j < n2) {
            pushEvent(events, arr, Arrays.asList(left + i, mid + 1 + j), "comparison", "compare", "Comparing elements", 10, visualization);
            if (L[i] <= R[j]) {
                arr[k] = L[i];
                i++;
            } else {
                arr[k] = R[j];
                j++;
            }
            pushEvent(events, arr, Arrays.asList(k), "array_mutation", "merge", "Merging element back", 10, visualization);
            k++;
        }
        while (i < n1) {
            arr[k] = L[i];
            pushEvent(events, arr, Arrays.asList(k), "array_mutation", "merge", "Merging remaining left", 13, visualization);
            i++;
            k++;
        }
        while (j < n2) {
            arr[k] = R[j];
            pushEvent(events, arr, Arrays.asList(k), "array_mutation", "merge", "Merging remaining right", 14, visualization);
            j++;
            k++;
        }
    }

    public static AlgorithmResult runInsertionSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Insertion Sort", 1, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        for (int i = 1; i < n; ++i) {
            int key = arr[i];
            int j = i - 1;
            pushEvent(events, arr, Arrays.asList(i), "comparison", "pick", "Picking element", 3, visualization);
            
            while (j >= 0 && arr[j] > key) {
                pushEvent(events, arr, Arrays.asList(j, j + 1), "comparison", "compare", "Comparing elements", 5, visualization);
                arr[j + 1] = arr[j];
                pushEvent(events, arr, Arrays.asList(j, j + 1), "array_mutation", "shift", "Shifting element", 6, visualization);
                j = j - 1;
            }
            arr[j + 1] = key;
            pushEvent(events, arr, Arrays.asList(j + 1), "array_mutation", "insert", "Inserted element", 8, visualization);
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Insertion Sort complete", 10, visualization);
        return new AlgorithmResult("insertion_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    public static AlgorithmResult runSelectionSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Selection Sort", 1, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            int min_idx = i;
            for (int j = i + 1; j < n; j++) {
                pushEvent(events, arr, Arrays.asList(min_idx, j), "comparison", "compare", "Comparing elements", 6, visualization);
                if (arr[j] < arr[min_idx]) {
                    min_idx = j;
                }
            }
            if (min_idx != i) {
                int temp = arr[min_idx];
                arr[min_idx] = arr[i];
                arr[i] = temp;
                pushEvent(events, arr, Arrays.asList(i, min_idx), "array_mutation", "swap", "Swapped elements", 9, visualization);
            }
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Selection Sort complete", 12, visualization);
        return new AlgorithmResult("selection_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    public static AlgorithmResult runHeapSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Heap Sort", 13, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arr, n, i, events, visualization);
        }
        
        for (int i = n - 1; i > 0; i--) {
            pushEvent(events, arr, Arrays.asList(0, i), "comparison", "pick", "Picking root", 17, visualization);
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            pushEvent(events, arr, Arrays.asList(0, i), "array_mutation", "swap", "Swapped root", 19, visualization);
            heapify(arr, i, 0, events, visualization);
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Heap Sort complete", 24, visualization);
        return new AlgorithmResult("heap_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    private static void heapify(int[] arr, int n, int i, List<Event> events, boolean visualization) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        
        if (l < n) {
            pushEvent(events, arr, Arrays.asList(largest, l), "comparison", "compare", "Comparing with left child", 3, visualization);
            if (arr[l] > arr[largest]) largest = l;
        }
        if (r < n) {
            pushEvent(events, arr, Arrays.asList(largest, r), "comparison", "compare", "Comparing with right child", 4, visualization);
            if (arr[r] > arr[largest]) largest = r;
        }
        
        if (largest != i) {
            pushEvent(events, arr, Arrays.asList(i, largest), "comparison", "pick", "Swapping child", 6, visualization);
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
            pushEvent(events, arr, Arrays.asList(i, largest), "array_mutation", "swap", "Swapped elements", 7, visualization);
            heapify(arr, n, largest, events, visualization);
        }
    }

    public static AlgorithmResult runRadixSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Radix Sort", 1, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        if (n > 0) {
            int max = arr[0];
            for (int i = 1; i < n; i++) {
                if (arr[i] > max) max = arr[i];
            }
            
            for (int exp = 1; max / exp > 0; exp *= 10) {
                countSort(arr, n, exp, events, visualization);
            }
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Radix Sort complete", 20, visualization);
        return new AlgorithmResult("radix_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }

    private static void countSort(int[] arr, int n, int exp, List<Event> events, boolean visualization) {
        int[] output = new int[n];
        int[] count = new int[10];
        
        for (int i = 0; i < n; i++) {
            pushEvent(events, arr, Arrays.asList(i), "comparison", "count", "Counting digit", 4, visualization);
            count[(arr[i] / exp) % 10]++;
        }
        
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }
        
        for (int i = n - 1; i >= 0; i--) {
            int idx = (arr[i] / exp) % 10;
            output[count[idx] - 1] = arr[i];
            count[idx]--;
        }
        
        for (int i = 0; i < n; i++) {
            arr[i] = output[i];
            pushEvent(events, arr, Arrays.asList(i), "array_mutation", "place", "Placed element", 15, visualization);
        }
    }

    public static AlgorithmResult runShellSort(int[] arr, boolean visualization) {
        List<Event> events = new ArrayList<>();
        pushEvent(events, arr, new ArrayList<>(), "initial", "start", "Starting Shell Sort", 1, visualization);
        long start = System.nanoTime();
        
        int n = arr.length;
        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j = i;
                while (j >= gap) {
                    pushEvent(events, arr, Arrays.asList(j - gap, i), "comparison", "compare", "Comparing elements", 6, visualization);
                    if (arr[j - gap] > temp) {
                        arr[j] = arr[j - gap];
                        pushEvent(events, arr, Arrays.asList(j), "array_mutation", "shift", "Shifting element", 7, visualization);
                        j -= gap;
                    } else {
                        break;
                    }
                }
                arr[j] = temp;
                pushEvent(events, arr, Arrays.asList(j), "array_mutation", "insert", "Inserted element", 10, visualization);
            }
        }
        
        long end = System.nanoTime();
        pushEvent(events, arr, new ArrayList<>(), "final", "end", "Shell Sort complete", 13, visualization);
        return new AlgorithmResult("shell_sort", visualization ? "visualize" : "benchmark", events, (end - start) / 1_000_000.0);
    }
}

File: rust-service/src/algorithms.rs
use crate::contracts::*;
use std::time::{Instant, SystemTime, UNIX_EPOCH};

fn get_timestamp() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn push_event(
    events: &mut Vec<AlgorithmEvent>,
    category: EventCategory,
    event: &str,
    indices: Option<Vec<usize>>,
    values: Option<Vec<i32>>,
    description: &str,
    metadata: Option<serde_json::Value>,
) {
    events.push(AlgorithmEvent {
        timestamp: get_timestamp(),
        category,
        event: event.to_string(),
        indices,
        values,
        description: Some(description.to_string()),
        metadata,
    });
}

fn line_meta(line: u32) -> Option<serde_json::Value> {
    Some(serde_json::json!({ "line_number": line }))
}

pub fn get_complexity(algo: &str) -> AlgorithmComplexity {
    match algo {
        "bubble_sort" | "insertion_sort" | "selection_sort" => AlgorithmComplexity {
            time: "O(n^2)".to_string(),
            space: "O(1)".to_string(),
        },
        "quick_sort" | "merge_sort" | "heap_sort" | "shell_sort" => AlgorithmComplexity {
            time: "O(n log n)".to_string(),
            space: "O(log n)".to_string(),
        },
        _ => AlgorithmComplexity {
            time: "O(?)".to_string(),
            space: "O(?)".to_string(),
        },
    }
}

fn bubble_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    let n = data.len();
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }

    for i in 0..n {
        for j in 0..n.saturating_sub(i).saturating_sub(1) {
            *comps += 1;
            if matches!(mode, AlgorithmMode::Visualization) {
                push_event(
                    events,
                    EventCategory::Comparison,
                    "compare",
                    Some(vec![j, j + 1]),
                    Some(vec![data[j], data[j + 1]]),
                    &format!("Comparing {} and {}", data[j], data[j + 1]),
                    line_meta(5),
                );
            }
            if data[j] > data[j + 1] {
                data.swap(j, j + 1);
                *swaps += 1;
                if matches!(mode, AlgorithmMode::Visualization) {
                    push_event(
                        events,
                        EventCategory::ArrayMutation,
                        "swap",
                        Some(vec![j, j + 1]),
                        Some(vec![data[j], data[j + 1]]),
                        &format!("Swapping {} and {}", data[j], data[j + 1]),
                        line_meta(6),
                    );
                }
            }
        }
    }

    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

fn quick_sort_helper(data: &mut Vec<i32>, low: usize, high: usize, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if low < high {
        let pivot_idx = partition(data, low, high, mode, events, comps, swaps);
        if pivot_idx > 0 {
            quick_sort_helper(data, low, pivot_idx - 1, mode, events, comps, swaps);
        }
        quick_sort_helper(data, pivot_idx + 1, high, mode, events, comps, swaps);
    }
}

fn partition(data: &mut Vec<i32>, low: usize, high: usize, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) -> usize {
    let pivot = data[high];
    let mut i = low;

    for j in low..high {
        *comps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::Comparison,
                "compare",
                Some(vec![j, high]),
                Some(vec![data[j], pivot]),
                &format!("Comparing {} with pivot {}", data[j], pivot),
                line_meta(5),
            );
        }
        
        if data[j] < pivot {
            if i != j {
                data.swap(i, j);
                *swaps += 1;
                if matches!(mode, AlgorithmMode::Visualization) {
                    push_event(
                        events,
                        EventCategory::ArrayMutation,
                        "swap",
                        Some(vec![i, j]),
                        Some(vec![data[i], data[j]]),
                        &format!("Swapping {} and {}", data[i], data[j]),
                        line_meta(6),
                    );
                }
            }
            i += 1;
        }
    }
    
    if i != high {
        data.swap(i, high);
        *swaps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::ArrayMutation,
                "swap_pivot",
                Some(vec![i, high]),
                Some(vec![data[i], data[high]]),
                &format!("Placing pivot {} at correct position", data[i]),
                line_meta(10),
            );
        }
    }
    
    i
}

fn quick_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if !data.is_empty() {
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
        }
        let n = data.len();
        quick_sort_helper(data, 0, n - 1, mode, events, comps, swaps);
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
        }
    }
}

fn merge_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if data.is_empty() { return; }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }
    let n = data.len();
    merge_sort_helper(data, 0, n - 1, mode, events, comps, swaps);
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

fn merge_sort_helper(data: &mut Vec<i32>, left: usize, right: usize, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if left < right {
        let mid = left + (right - left) / 2;
        merge_sort_helper(data, left, mid, mode, events, comps, swaps);
        merge_sort_helper(data, mid + 1, right, mode, events, comps, swaps);
        merge(data, left, mid, right, mode, events, comps, swaps);
    }
}

fn merge(data: &mut Vec<i32>, left: usize, mid: usize, right: usize, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    let mut temp = Vec::with_capacity(right - left + 1);
    let mut i = left;
    let mut j = mid + 1;

    while i <= mid && j <= right {
        *comps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::Comparison,
                "compare",
                Some(vec![i, j]),
                Some(vec![data[i], data[j]]),
                &format!("Comparing {} and {}", data[i], data[j]),
                line_meta(7),
            );
        }
        if data[i] <= data[j] {
            temp.push(data[i]);
            i += 1;
        } else {
            temp.push(data[j]);
            j += 1;
        }
    }

    while i <= mid {
        temp.push(data[i]);
        i += 1;
    }

    while j <= right {
        temp.push(data[j]);
        j += 1;
    }

    for (k, &val) in temp.iter().enumerate() {
        let idx = left + k;
        data[idx] = val;
        *swaps += 1; // treating write as swap for metric parity with other languages
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::ArrayMutation,
                "overwrite",
                Some(vec![idx]),
                Some(vec![val]),
                &format!("Merging value {} back to array", val),
                line_meta(27),
            );
        }
    }
}

fn insertion_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }
    let n = data.len();
    for i in 1..n {
        let key = data[i];
        let mut j = i;
        
        while j > 0 {
            *comps += 1;
            if matches!(mode, AlgorithmMode::Visualization) {
                push_event(
                    events,
                    EventCategory::Comparison,
                    "compare",
                    Some(vec![j - 1, i]),
                    Some(vec![data[j - 1], key]),
                    &format!("Comparing {} with key {}", data[j - 1], key),
                    line_meta(7),
                );
            }
            if data[j - 1] > key {
                data[j] = data[j - 1];
                *swaps += 1;
                if matches!(mode, AlgorithmMode::Visualization) {
                    push_event(
                        events,
                        EventCategory::ArrayMutation,
                        "shift",
                        Some(vec![j]),
                        Some(vec![data[j]]),
                        &format!("Shifting {} to the right", data[j]),
                        line_meta(8),
                    );
                }
                j -= 1;
            } else {
                break;
            }
        }
        data[j] = key;
        *swaps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::ArrayMutation,
                "insert",
                Some(vec![j]),
                Some(vec![data[j]]),
                &format!("Inserting key {} at position {}", key, j),
                line_meta(11),
            );
        }
    }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

fn selection_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }
    let n = data.len();
    for i in 0..n {
        let mut min_idx = i;
        for j in (i + 1)..n {
            *comps += 1;
            if matches!(mode, AlgorithmMode::Visualization) {
                push_event(
                    events,
                    EventCategory::Comparison,
                    "compare",
                    Some(vec![j, min_idx]),
                    Some(vec![data[j], data[min_idx]]),
                    &format!("Comparing {} with current min {}", data[j], data[min_idx]),
                    line_meta(6),
                );
            }
            if data[j] < data[min_idx] {
                min_idx = j;
            }
        }
        if min_idx != i {
            data.swap(i, min_idx);
            *swaps += 1;
            if matches!(mode, AlgorithmMode::Visualization) {
                push_event(
                    events,
                    EventCategory::ArrayMutation,
                    "swap",
                    Some(vec![i, min_idx]),
                    Some(vec![data[i], data[min_idx]]),
                    &format!("Swapping {} with new minimum {}", data[i], data[min_idx]),
                    line_meta(11),
                );
            }
        }
    }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

fn heap_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if data.is_empty() { return; }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }
    let n = data.len();

    // Build max heap
    for i in (0..n / 2).rev() {
        heapify(data, n, i, mode, events, comps, swaps);
    }

    // Extract elements from heap one by one
    for i in (1..n).rev() {
        data.swap(0, i);
        *swaps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::ArrayMutation,
                "swap",
                Some(vec![0, i]),
                Some(vec![data[0], data[i]]),
                &format!("Moving root {} to end of heap", data[i]),
                line_meta(29),
            );
        }
        heapify(data, i, 0, mode, events, comps, swaps);
    }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

fn heapify(data: &mut Vec<i32>, n: usize, i: usize, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    let mut largest = i;
    let left = 2 * i + 1;
    let right = 2 * i + 2;

    if left < n {
        *comps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::Comparison,
                "compare",
                Some(vec![left, largest]),
                Some(vec![data[left], data[largest]]),
                &format!("Comparing child {} with parent {}", data[left], data[largest]),
                line_meta(6),
            );
        }
        if data[left] > data[largest] {
            largest = left;
        }
    }

    if right < n {
        *comps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::Comparison,
                "compare",
                Some(vec![right, largest]),
                Some(vec![data[right], data[largest]]),
                &format!("Comparing child {} with parent {}", data[right], data[largest]),
                line_meta(10),
            );
        }
        if data[right] > data[largest] {
            largest = right;
        }
    }

    if largest != i {
        data.swap(i, largest);
        *swaps += 1;
        if matches!(mode, AlgorithmMode::Visualization) {
            push_event(
                events,
                EventCategory::ArrayMutation,
                "swap",
                Some(vec![i, largest]),
                Some(vec![data[i], data[largest]]),
                &format!("Swapping parent {} with child {}", data[largest], data[i]),
                line_meta(15),
            );
        }
        heapify(data, n, largest, mode, events, comps, swaps);
    }
}

fn shell_sort(data: &mut Vec<i32>, mode: &AlgorithmMode, events: &mut Vec<AlgorithmEvent>, comps: &mut u64, swaps: &mut u64) {
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Initial, "initial", None, Some(data.clone()), "Initial state", None);
    }
    let n = data.len();
    let mut gap = n / 2;

    while gap > 0 {
        for i in gap..n {
            let temp = data[i];
            let mut j = i;

            while j >= gap {
                *comps += 1;
                if matches!(mode, AlgorithmMode::Visualization) {
                    push_event(
                        events,
                        EventCategory::Comparison,
                        "compare",
                        Some(vec![j - gap, i]),
                        Some(vec![data[j - gap], temp]),
                        &format!("Comparing elements at indices {} and {} with gap {}", j - gap, i, gap),
                        line_meta(10),
                    );
                }
                if data[j - gap] > temp {
                    data[j] = data[j - gap];
                    *swaps += 1;
                    if matches!(mode, AlgorithmMode::Visualization) {
                        push_event(
                            events,
                            EventCategory::ArrayMutation,
                            "shift",
                            Some(vec![j]),
                            Some(vec![data[j]]),
                            &format!("Shifting element {} forward", data[j]),
                            line_meta(11),
                        );
                    }
                    j -= gap;
                } else {
                    break;
                }
            }
            data[j] = temp;
            *swaps += 1;
            if matches!(mode, AlgorithmMode::Visualization) {
                push_event(
                    events,
                    EventCategory::ArrayMutation,
                    "insert",
                    Some(vec![j]),
                    Some(vec![data[j]]),
                    &format!("Inserting {} at correct gap position", data[j]),
                    line_meta(14),
                );
            }
        }
        gap /= 2;
    }
    if matches!(mode, AlgorithmMode::Visualization) {
        push_event(events, EventCategory::Final, "final", None, Some(data.clone()), "Sorted state", None);
    }
}

pub fn run_sort(algo: &str, mut data: Vec<i32>, mode: &AlgorithmMode) -> VersionedAlgorithmContract {
    let mut events = Vec::new();
    let mut comparisons = 0;
    let mut swaps = 0;
    
    let start = Instant::now();
    
    match algo {
        "bubble_sort" => bubble_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "quick_sort" => quick_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "merge_sort" => merge_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "insertion_sort" => insertion_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "selection_sort" => selection_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "heap_sort" => heap_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        "shell_sort" => shell_sort(&mut data, mode, &mut events, &mut comparisons, &mut swaps),
        _ => {
            // Fallback default sort
            data.sort();
        }
    }
    
    let duration = start.elapsed();
    let time_ms = duration.as_secs_f64() * 1000.0;
    
    let final_state = match mode {
        AlgorithmMode::Visualization => Some(data.iter().map(|&v| serde_json::Value::Number(v.into())).collect()),
        AlgorithmMode::Benchmark => None,
    };
    
    VersionedAlgorithmContract {
        version: "1.0".to_string(),
        algorithm: algo.to_string(),
        language: "rust".to_string(),
        mode: mode.clone(),
        visualization_level: None,
        events,
        metrics: AlgorithmMetrics {
            comparisons: Some(comparisons),
            swaps: Some(swaps),
            time_ms,
            allocations: None,
            memory_used_kb: None,
        },
        complexity: get_complexity(algo),
        final_state,
    }
}
